File: 0909-snakes-and-ladders/0909-snakes-and-ladders.cpp
class Solution {
public:
    void getCoordinates(int n, int pos, int& row, int& col) {
        row = n - 1 - (pos - 1) / n;
        col = (pos - 1) % n;
        if ((n % 2 == 1 && row % 2 == 1) || (n % 2 == 0 && row % 2 == 0)) {
            col = n - 1 - col;
        }
    }
    
    int snakesAndLadders(vector<vector<int>>& board) {
        int n = board.size();
        queue<pair<int, int>> q; // { square position, number of moves }
        vector<bool> visited(n * n + 1, false);
        q.push({ 1, 0 });
        visited[1] = true;
        while (!q.empty()) {
            int currPos = q.front().first;
            int moves = q.front().second;
            q.pop();
            if (currPos == n * n) {
                return moves;
            }
            for (int next = currPos + 1; next <= min(n * n, currPos + 6); next++) {
                int row, col;
                getCoordinates(n, next, row, col);
                int newPos = (board[row][col] == -1) ? next : board[row][col];
                if (!visited[newPos]) {
                    q.push({ newPos, moves + 1 });
                    visited[newPos] = true;
                }
            }
        }
        return -1;
    }
};

File: 0909-snakes-and-ladders/0909-snakes-and-ladders.java
class Solution {
    public int snakesAndLadders(int[][] board) {
        int n = board.length, moves = 0;
        int dest = n * n;
        Queue<Integer> queue = new LinkedList<>();
        queue.offer(1);
        boolean[] visited = new boolean[dest + 1];
        visited[1] = true;
        while (!queue.isEmpty()) {
            int size = queue.size();
            moves += 1;
            while (size-- > 0) {
                int curr = queue.poll();
                for (int dice = 1; dice <= 6; dice++) {
                    int next = Math.min(curr + dice, dest);
                    if (visited[next]) {
                        continue;
                    }
                    visited[next] = true;
                    int row = n - 1 - (next - 1) / n;
                    int col = (next - 1) % n;
                    if ((n % 2) == (row % 2)) {
                        col = n - 1 - col;
                    }
                    if (board[row][col] != -1) {
                        next = board[row][col];
                    }
                    if (next == dest) {
                        return moves;
                    }
                    queue.offer(next);
                }
            }
        }
        return -1;
    }
}